src/Neo/Extensions/NeoSystemExtensions.cs

```diff
@@ -0,0 +1,99 @@
+// Copyright (C) 2015-2025 The Neo Project.
+//
+// NeoSystemExtensions.cs file belongs to the neo project and is free
+// software distributed under the MIT software license, see the
+// accompanying file LICENSE in the main directory of the
+// repository or http://www.opensource.org/licenses/mit-license.php
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+/* Unmerged change from project 'Neo(net9.0)'
+Before:
+using System;
+After:
+// Copyright (C) 2015-2025 The Neo Project.
+//
+// NeoSystemExtensions.cs file belongs to the neo project and is free
+// software distributed under the MIT software license, see the
+// accompanying file LICENSE in the main directory of the
+// repository or http://www.opensource.org/licenses/mit-license.php
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+using System;
+*/
+// Copyright (C) 2015-2025 The Neo Project.
+//
+// NeoSystemExtensions.cs file belongs to the neo project and is free
+// software distributed under the MIT software license, see the
+// accompanying file LICENSE in the main directory of the
+// repository or http://www.opensource.org/licenses/mit-license.php
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+using Neo.Network.P2P.Payloads;
+using Neo.Persistence;
+using Neo.SmartContract.Native;
+using System;
+
+namespace Neo
+{
+    /// <summary>
+    /// Extension methods for <see cref="NeoSystem"/> and related types.
+    /// </summary>
+    public static class NeoSystemExtensions
+    {
+        /// <summary>
+        /// Gets the block generation time based on the current state of the blockchain.
+        /// </summary>
+        /// <param name="system">The NeoSystem instance.</param>
+        /// <returns>The block generation time as a TimeSpan.</returns>
+        public static TimeSpan GetBlockGenTime(this NeoSystem system)
+        {
+            try
+            {
+                // Get the current block height from the blockchain
+                var index = NativeContract.Ledger.CurrentIndex(system.StoreView);
+
+                // Before the Echidna hardfork, use the protocol settings
+                if (!system.Settings.IsHardforkEnabled(Hardfork.HF_Echidna, index - 1))
+                    return TimeSpan.FromMilliseconds(system.Settings.MillisecondsPerBlock);
+
+                // After the Echidna hardfork, get the current block time from the Policy contract
+                var milliseconds = NativeContract.Policy.GetBlockGenTime(system.StoreView);
+                return TimeSpan.FromMilliseconds(milliseconds);
```

Comment:
We can reuse `GetBlockGenTime(this IReadOnlyStore snapshot, ProtocolSettings settings)` extension from this method in order not to duplicate the code. Can't we?

Reply:
but the execution of contract are after consensus, which means if height 1000 enables we can read the value at height 1001. 

this change is different from all other hardfork change cause previously all changes take effect at block runtime, while this change take effect at the consensus time, which is before the block runtime.